### PR TITLE
feat(cli): better image pull display

### PR DESCRIPTION
**Pull request by nandor-magyar**
Done:
- imagehelper with customImagePull function that has a callback for display
- CLI has a display function implemented
- removed dead code

### CODE REVIEW

**Comment by nandor-magyar** [comment]
Something is messed up with git, gonna fix soon.

File: golang/pkg/builder/container/container_builder.go [hunk 1]
```diff
@@ -501,26 +503,27 @@ func execHooks(dc *DockerContainerBuilder, hooks []LifecycleFunc) error {
...
+// 	if imageExists {
+// 		return false, nil
+// 	}
+
+// 	imageExists, err = imageHelper.Exists(ctx, logger, ref.String())
+// 	if err != nil {
+// 		return false, err
+// 	}
+// 	return !imageExists, nil
+// }
```
Comment (polaroi8d) [inline golang/pkg/builder/container/container_builder.go:526]: ?

File: golang/pkg/builder/container/container_builder.go [hunk 2]
```diff
@@ -501,26 +503,27 @@ func execHooks(dc *DockerContainerBuilder, hooks []LifecycleFunc) error {
...
-		return false, nil
-	}
-
-	imageExists, err = imageHelper.Exists(ctx, logger, ref.String())
-	if err != nil {
-		return false, err
-	}
-	return !imageExists, nil
-}
+// checked by the new pull
```
Comment (robot9706) [inline golang/pkg/builder/container/container_builder.go:506]: Comment? :)

File: golang/internal/helper/image/image.go [hunk 3]
```diff
@@ -119,7 +117,7 @@ func Exists(ctx context.Context, logger io.StringWriter, expandedImageName strin
 }
 
 // force pulls the given image name
-// TODO(nandor-magyar): the output from docker is not really nice, should be improved
+// todo(nandor-magyar): the output from docker is not really nice, should be improved
```
Comment (robot9706) [inline golang/internal/helper/image/image.go:120]: We use upper case TODO, no?

**Review by robot9706 — COMMENTED** [review]
For some reason image pulling doesn't work.
I deleted all images locally so I could see the pretty logging, but CLI won't pull any images.

I think it has something to do with image names as `docker.io/library/traefik:v2.9` does exist locally, but the CLI tries to start `traefik:v2.9` which doesn't work. CLI on `develop` works though.

```
9:09AM INF Pull docker.io/library/traefik:v2.9 status: ✓ up-to-date
Container network: [1907a4e701388a93e536a9d4b22a902b3abb0cd3e3a00201834c91ad56d5337b]
Container create failed:  Error response from daemon: No such image: traefik:v2.9

Container was not created.
9:09AM ERR Failed to start dyrector io stack container=traefik
9:09AM FTL error="container was not created"
```

File: golang/internal/helper/image/image.go (hunk 3 above)
Comment (polaroi8d) [reply golang/internal/helper/image/image.go:120]: Yeah, with a `@`

File: golang/pkg/cli/display.go [hunk 4]
```diff
@@ -48,3 +55,62 @@ func NotifyOnce(name string, notifyFunc func()) {
...
+				log.Info().Msgf("%v %s layers: %d/%d", header, spinner(i), pulled, len(stat))
+				tm.MoveCursorUp(1)
+				tm.Flush()
+			}
+		}
+	}
+	return nil
+}
+
+var spinChars = []string{"⢹", "⢺", "⢼", "⣸", "⣇", "⡧", "⡗", "⡏"}
```
Comment (polaroi8d) [inline golang/pkg/cli/display.go:112]: `spinChars := []string{"⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"}`

**Comment by m8vago** [comment]
@nandor-magyar updated the branch to the latest dev

File: golang/pkg/builder/container/container_builder.go (hunk 1 above)
Comment (nandor-magyar) [reply golang/pkg/builder/container/container_builder.go]: Removed as it is unnecessary now.

File: golang/pkg/builder/container/container_builder.go (hunk 2 above)
Comment (nandor-magyar) [reply golang/pkg/builder/container/container_builder.go]: Gone.

File: golang/pkg/cli/cli.go [hunk 5]
```diff
@@ -9,6 +9,23 @@ import (
...
+	FlagDisableCrux           = "disable-crux"
+	FlagDisableCruxUI         = "disable-crux-ui"
+	FlagLocalAgent            = "local-agent"
+	FlagWrite                 = "write"
+	FlagDebug                 = "debug"
+	FlagPrioritizeLocalImages = "prioritize-local-images"
+	FlagDisablePodmanChecks   = "disable-podman-checks"
+	FlagConfigPath            = "config"
+	FlagPrefix                = "prefix"
+	FlagImageTag              = "image-tag"
```
Comment (robot9706) [inline golang/pkg/cli/cli.go:23]: Makefile upd uses `imagetag` instead of `image-tag` :)

Comment (nandor-magyar) [reply golang/pkg/cli/cli.go:23]: Yes, good catch, found some additional leftovers too.

File: golang/pkg/cli/runner.go [hunk 6]
```diff
@@ -88,6 +90,16 @@ func ProcessCommand(ctx context.Context, initialState *State, args *ArgsFlags) {
...
+	case VersionCommand:
+		cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+		if err != nil {
+			log.Fatal().Err(err).Msg("Could not connect to docker socket.")
+		}
+		out, err := container.VersionCheck(ctx, cli)
+		if err != nil {
+			log.Fatal().Err(err).Msg("Version error")
+		}
+		out.Msg("")
```
Comment (polaroi8d) [inline golang/pkg/cli/runner.go:102]: Why need an empty message?

File: golang/pkg/cli/display.go [hunk 7]
```diff
@@ -52,3 +59,61 @@ func NotifyOnce(name string, notifyFunc func()) {
...
+			phase == imageHelper.LayerProgressStatusWaiting:
+			stat[jm.ID].Total = jm.Progress.Total
+			waiting++
+		case phase == imageHelper.LayerProgressStatusDownloading:
+			stat[jm.ID].Current = jm.Progress.Current
+			pulling++
+		case phase == imageHelper.LayerProgressStatusComplete || phase == imageHelper.LayerProgressStatusExists:
+			pulled++
+		}
+		if phase != imageHelper.LayerProgressStatusUnknown {
```
Comment (polaroi8d) [inline golang/pkg/cli/display.go:105]: Why not 
```if phase != imageHelper.LayerProgressStatusUnknown && len(stat) > 1 {
```?

File: golang/pkg/cli/display.go [hunk 8]
```diff
@@ -52,3 +59,61 @@ func NotifyOnce(name string, notifyFunc func()) {
...
+}
+
+func DockerPullProgressDisplayer(header string, respIn io.ReadCloser) error {
+	if respIn == nil {
+		log.Info().Msgf("%s ✓ up-to-date", header)
+		return nil
+	}
+
+	dec := json.NewDecoder(respIn)
+	stat := map[string]*status{}
```
Comment (polaroi8d) [inline golang/pkg/cli/display.go:75]: I'm not a go hero, which is the preferred way? 

```stat := make(map[string]*status)```

**Comment by polaroi8d** [comment]
Tested with pure system. Got the following when used `make upd` 

```
❯ make upd
cd golang && \
	make upd
cd cmd/dyo && \
	go run . --disable-crux --disable-crux-ui --image-tag latest --prefix dyo-latest up
1:22PM INF There is a newer version of the container engine in use, please consider updating. error="server is outdated"
1:22PM INF Pull docker.io/library/traefik:v2.9 status: ✓ pull complete
1:22PM ERR Failed to start dyrector.io stack container=traefik
1:22PM FTL error="container was not created"
```

**Review by robot9706 — COMMENTED** [review]
Getting the same error as @polaroi8d:

```
go run . --disable-crux --disable-crux-ui --image-tag latest --prefix dyo-latest up
11:24AM INF There is a newer version of the container engine in use, please consider updating. error="server is outdated"
11:24AM INF Pull docker.io/library/traefik:v2.9 status: ✓ pull complete 
11:24AM ERR Failed to start dyrector.io stack container=traefik
11:24AM FTL error="container was not created"
exit status 1
```

I think it might have to do something with image names, while pull works for "docker.io/library/traefik:v2.9" starting a container doesn't work with "traefik:v2.9" or something like that.

File: golang/pkg/cli/display.go (hunk 8 above)
Comment (nandor-magyar) [reply golang/pkg/cli/display.go:75]: I don't like make, they are equivalent.

File: golang/internal/runtime/container/container_test.go [hunk 9]
```diff
@@ -54,34 +54,80 @@ func getDockerInfoInvalid() *types.Info {
...
+			Info:              getDockerInfoInvalid(),
+			MockClientVersion: "23.0.0",
+			ErrExpected:       containerRuntime.ErrServerUnknown,
+		},
+		{
+			Info:              getDockerInfoInvalid(),
+			MockClientVersion: "4.4.0",
+			ErrExpected:       containerRuntime.ErrServerUnknown,
+		},
+		// We don't check the specific error there, since it is not covered by the pakcage but by an external dependency.
```
Comment (robot9706) [inline golang/internal/runtime/container/container_test.go:109]: `by the pakcage but by`
typo: package

File: golang/pkg/cli/runner.go (hunk 6 above)
Comment (nandor-magyar) [reply golang/pkg/cli/runner.go:102]: Something to terminate / print the zerolog event must be called for it to be visible.

File: golang/pkg/cli/display.go (hunk 7 above)
Comment (nandor-magyar) [reply golang/pkg/cli/display.go:105]: Merged.

File: golang/internal/runtime/container/container_test.go (hunk 9 above)
Comment (nandor-magyar) [reply golang/internal/runtime/container/container_test.go]: corrected

File: golang/pkg/cli/config_file.go [hunk 10]
```diff
@@ -278,10 +277,10 @@ func LoadDefaultsOnEmpty(state *State, args *ArgsFlags) *State {
 	state.Kratos.Image = "ghcr.io/dyrector-io/dyrectorio/web/kratos"
 
 	// Load defaults
-	state.SettingsFile.CruxSecret = util.Fallback(state.SettingsFile.CruxSecret, randomChars())
-	state.SettingsFile.CruxPostgresPassword = util.Fallback(state.SettingsFile.CruxPostgresPassword, randomChars())
-	state.SettingsFile.KratosPostgresPassword = util.Fallback(state.SettingsFile.KratosPostgresPassword, randomChars())
-	state.SettingsFile.KratosSecret = util.Fallback(state.SettingsFile.KratosSecret, randomChars())
+	state.SettingsFile.CruxSecret = util.Fallback(state.SettingsFile.CruxSecret, RandomChars(secretLength))
```
Comment (c3ppc3pp) [inline golang/pkg/cli/config_file.go:280]: Any reason to revert these?

File: golang/pkg/cli/runner.go (hunk 6 above)
Comment (c3ppc3pp) [reply golang/pkg/cli/runner.go:102]: `Send()` ?

**Review by polaroi8d — APPROVED** [review]
Works 🥳 Thanks!

File: golang/pkg/cli/runner.go (hunk 6 above)
Comment (nandor-magyar) [reply golang/pkg/cli/runner.go:102]: We can agree on using not using Msg("") or Msgf("") but then it must be a lint rule too.

File: golang/pkg/cli/config_file.go (hunk 10 above)
Comment (nandor-magyar) [reply golang/pkg/cli/config_file.go]: Mistakes were made.

**Pull request merged**